mount/mount_linux.go

```diff
@@ -117,9 +193,45 @@ func (m *Mount) mount(target string) (err error) {
 		// Remount the bind to apply read only.
 		return unix.Mount("", target, "", uintptr(oflags|unix.MS_REMOUNT), "")
 	}
+
+	// remap non-overlay mount point
+	if opt.uidmap != "" && opt.gidmap != "" && m.Type != "overlay" {
+		if err := IDMapMount(target, target, usernsFd); err != nil {
+			return err
+		}
+	}
 	return nil
 }
 
+func doPrepareIDMappedOverlay(lowerDirs []string, usernsFd int) (tmpLowerDirs []string, _ func(), _ error) {
+	td, err := os.MkdirTemp(tempMountLocation, "ovl-idmapped")
+	if err != nil {
+		return nil, nil, err
+	}
+	cleanUp := func() {
+		for _, lowerDir := range tmpLowerDirs {
+			if err := unix.Unmount(lowerDir, 0); err != nil {
+				logrus.WithError(err).Warnf("failed to unmount temp lowerdir %s", lowerDir)
+			}
+		}
+		if terr := os.RemoveAll(filepath.Clean(filepath.Join(tmpLowerDirs[0], ".."))); terr != nil {
+			logrus.WithError(terr).Warnf("failed to remove temporary overlay lowerdir's")
+		}
+	}
+	for i, lowerDir := range lowerDirs {
```

Comment:
You only idmap mount the lower dir. I don't know, but is upper or work dir not needed to be idmapped too? Will it work fine if we create a snapshot (IIRC that stores the upper dir, maybe the work dir too) and restore it in a container using different userns mappings?

I'm trying to find out if we can idmap the upper/work dir too, but I can't make it work locally, it fails with:

```
overlayfs: workdir and upperdir must reside under the same mount
```

Whenever upper or work dir are idmapped.

Can you test a snapshot and see if restoring in a container with different mapping works fine?
Or do you know if we should or should not idmap the upper layer too?

Reply:
>IIRC that stores the upper dir, maybe the work dir too) and restore it in a container using different userns mappings

It should work fine, since we doesn't idmap upperdir, only `chown` it.

> overlayfs: workdir and upperdir must reside under the same mount

Yes, because `upperdir` and `workdir` have different mounts here, as far as I remember the implementation it creates new mount for every idmapped mount, so if we apply the same mapping, say in order then they are still different. 

>Can you test a snapshot and see if restoring in a container with different mapping works fine?
Or do you know if we should or should not idmap the upper layer too?

I'll try this. Also, it'd be better if you share how you exactly tested this. Thanks!

Reply:
Hmm... actually, you're right, there may be few issues about it, the first one I see is just running the same container after committing and preparing a snapshot of that one, I'm not sure how containerd does that, for my case I have only `/root` remapped. Not sure about the reason.

Actually, at the first glance I don't see proper solution for idmapped mounts for this, may be we need something like a "manifest" to be able to restore idmapped mappings or so. I vote for creating a follow-up issue for this.

Reply:
@IlyaHanov do you have the steps you did to try it handy? Can you share them here?

It would be nice if we chown the upper dir, at least, if that is simple

Reply:
Yeah, that's what I did (these steps don't include changing userns mapping):

```
# ctr i pull docker.io/library/ubuntu:latest
(output truncated)

# ctr run -t --uidmap="0:1000:65536" --gidmap="0:1000:65536" --remap-labels docker.io/library/ubuntu:latest ubuntu
root@host: /# exit

# ctr snapshots commit ubuntu_committed ubuntu
# ctr snapshots prepare ubuntu ubuntu_committed

# ctr t start ubuntu
root@host:/# ls -l
total 48
lrwxrwxrwx   1 nobody nogroup    7 Jun 24 02:02 bin -> usr/bin
drwxr-xr-x   2 nobody nogroup 4096 Apr 18  2022 boot
drwxr-xr-x   5 root   root     360 Aug  3 13:47 dev
drwxr-xr-x  32 nobody nogroup 4096 Jun 24 02:06 etc
drwxr-xr-x   2 nobody nogroup 4096 Apr 18  2022 home
lrwxrwxrwx   1 nobody nogroup    7 Jun 24 02:02 lib -> usr/lib
lrwxrwxrwx   1 nobody nogroup    9 Jun 24 02:02 lib32 -> usr/lib32
lrwxrwxrwx   1 nobody nogroup    9 Jun 24 02:02 lib64 -> usr/lib64
lrwxrwxrwx   1 nobody nogroup   10 Jun 24 02:02 libx32 -> usr/libx32
drwxr-xr-x   2 nobody nogroup 4096 Jun 24 02:02 media
drwxr-xr-x   2 nobody nogroup 4096 Jun 24 02:02 mnt
drwxr-xr-x   2 nobody nogroup 4096 Jun 24 02:02 opt
dr-xr-xr-x 382 nobody nogroup    0 Aug  3 13:47 proc
drwx------   1 root   root    4096 Aug  3 13:40 root
drwxr-xr-x   2 root   root      40 Aug  3 13:47 run
lrwxrwxrwx   1 nobody nogroup    8 Jun 24 02:02 sbin -> usr/sbin
drwxr-xr-x   2 nobody nogroup 4096 Jun 24 02:02 srv
dr-xr-xr-x  13 nobody nogroup    0 Aug  3 13:47 sys
drwxrwxrwt   2 nobody nogroup 4096 Jun 24 02:06 tmp
drwxr-xr-x  14 nobody nogroup 4096 Jun 24 02:02 usr
drwxr-xr-x  11 nobody nogroup 4096 Jun 24 02:06 var
root@host:/# 
```

And actually `rootfs` itself looks like:
```
lrwxrwxrwx  7 root     24 Jun 05:02 bin -> usr/bin
drwxr-xr-x  - root     18 Apr  2022 boot
drwxr-xr-x  - root     24 Jun 05:06 dev
drwxr-xr-x  - root     24 Jun 05:06 etc
drwxr-xr-x  - root     18 Apr  2022 home
lrwxrwxrwx  7 root     24 Jun 05:02 lib -> usr/lib
lrwxrwxrwx  9 root     24 Jun 05:02 lib32 -> usr/lib32
lrwxrwxrwx  9 root     24 Jun 05:02 lib64 -> usr/lib64
lrwxrwxrwx 10 root     24 Jun 05:02 libx32 -> usr/libx32
drwxr-xr-x  - root     24 Jun 05:02 media
drwxr-xr-x  - root     24 Jun 05:02 mnt
drwxr-xr-x  - root     24 Jun 05:02 opt
drwxr-xr-x  - root     18 Apr  2022 proc
drwx------  - 1000     3  Aug 16:40 root
drwxr-xr-x  - root     24 Jun 05:06 run
lrwxrwxrwx  8 root     24 Jun 05:02 sbin -> usr/sbin
drwxr-xr-x  - root     24 Jun 05:02 srv
drwxr-xr-x  - root     18 Apr  2022 sys
drwxrwxrwt  - root     24 Jun 05:06 tmp
drwxr-xr-x  - root     24 Jun 05:02 usr
drwxr-xr-x  - root     24 Jun 05:06 var
```

Test case about changing the userns mapping while keeping different idmapped mounts looks weird to me. Looks like `ctr` can't do this by default manipulating snaphots and containers, but it looks possible using `WithSnapshot`, still not sure, need time to check.

>It would be nice if we chown the upper dir, at least, if that is simple

We already do this, otherwise the owner of the rootfs directory itself won't be changed.

As I already said above we need to re-apply mapping for particular snapshot/container to be able to use it, but I'm not familiar with "commit" part of snapshotter, so not sure what it does then.

Reply:
>  but is upper or work dir not needed to be idmapped too?

The updir is empty at the beginning and the process has been forked into userns (root in userns -> N in host-userns). The write action will be translated to N in host-userns. You don't need to idmapped that.

Reply:
@IlyaHanov oh, no, I thought the example you had was "more real". I think it is ok if that case is not handled correctly.

@fuweid yes, I understand that. But isn't there any case where we can re-use a snapshot of a container with userns? I mean, if the upper layer is not idmapped, and we use the snapshot for other container, can't this be problematic? I'm not familiar with containerd snapshots, so far I couldn't find a scenario using `ctr`, but I thought there could be a use case we care about.

Do you know?

If such a use case exists, then having an idmap of the upper layer will completely solve it. Today the mount looks like this with this patchset:
```
overlay on / type overlay (rw,relatime,lowerdir=/tmp/ovl-idmapped2226919825/0,upperdir=/var/lib/containerd-rata/io.containerd.snapshotter.v1.overlayfs/snapshots/390/fs,workdir=/var/lib/containerd-rata/io.containerd.snapshotter.v1.overlayfs/snapshots/390/work)
``` 

So we can just idmap (in this example): `/var/lib/containerd-rata/io.containerd.snapshotter.v1.overlayfs/snapshots/390/` and it should just work, files will be created always with the same UID/GID. And the hierarchy is safe to assume that will be like this (work and fs dirs subdirs of a common dir), as work and fs should be in the same mount (that is enforced by the kernel).

But if such scenario can't exist, then this is perfectly fine :)

Reply:
For kubernetes, there is a case about checkpoint which might use the snapshot. And in some cases, the user can mount the containerd sock into pod and use nerdctl to commit the container as snapshot. For instance, the cluster admin provides container as vm to internal user. The user can shutdown the container and restart it. When they shutdown the container, the platform will commit the container as new image. And then after restart, the user can continue to work with last snapshot. In the case, we might run into the issue.

I think we can mark it as TODO or FIXME item and handle it in the follow-up. What do you think?

Reply:
@fuweid I think that seems okay if the only problems are those.

If you happen to have handy a way to run into those scenarios, it will be nice to document it in an issue so when we tackle we can verify we fix the issue.

I think the solution will be to idmap  work and upperdir params too, which hopefully shouldn't be too hard.

Reply:
> If you happen to have handy a way to run into those scenarios, it will be nice to document it in an issue so when we tackle we can verify we fix the issue

We can file issue to track this. This pull request should have fixed scope.

Reply:
Yes, I agree :)